### PR TITLE
feat: add annotations attribute to job spans

### PR DESCRIPTION
**Pull request by tcsullens**
Add span attribute for annotation messages from failed jobs to better attribute GHA job failures. This will allow us to collect and analyze root causes of failed jobs. 
This implementation collects annotations messages from only expected GitHub/Infra errors and joins each failure message into a `;`-separated string as the attribute value. We expect to first collect some data and then determine how best to setup an SLI to track/alert on GitHub/Infra failures.

**The actual code changes are in the `src/` folder**

An example using this test PR, https://github.com/wrapbook/app/pull/30740, sets a new span attribute `github.job.annotations` to the message of the annotation ([Trace Link](https://ui.honeycomb.io/wrapbook/environments/gha-ci/datasets/ci-v2/result/HxidhVCUrX3/trace/zTfLyzCYktg?fields[]=s_name&fields[]=s_serviceName&span=7e17ce3b047e3f76)):
Screenshot 2024-10-24 at 3 09 36 PM

Additional traces from testing: https://ui.honeycomb.io/wrapbook/environments/gha-ci/datasets/ci-v2/result/ANkTYWgHYXi?hideCompare

[sc-126750](https://app.shortcut.com/wrapbook/story/126750/collect-job-annotation-in-github-actions-job-otel-spans)

### CODE REVIEW

File: src/tracing/job.ts [hunk 1]
```diff
@@ -160,12 +162,28 @@ export async function traceWorkflowRunJobs({
...
+            check_run_id: job.id,
+          });
+        core.debug(`JobAnnotations Count: ${annotationsResponse.data.length}`);
+
+        // We only care about failure annotations generated by github files,
+        // excluding excluding annotations from other tools like rspec failures
+        const annotations = annotationsResponse.data.filter(
+          (a: JobAnnotation) =>
+            a.annotation_level === "failure" && a.path === ".github",
+        );
```
Comment (tcsullens) [inline src/tracing/job.ts:178]: Based on investigation of some historical jobs, GitHub and Infra-related failures originate from a `path` of `.github`, while an `rspec` failure originates from a `path` of the ruby spec file.

File: src/tracing/job.ts [hunk 2]
```diff
@@ -203,6 +223,10 @@ async function traceWorkflowRunJob({
   const ctx = trace.setSpan(parentContext, parentSpan);
   const startTime = new Date(job.started_at);
   const completedTime = new Date(job.completed_at);
+  const annotationMessages = jobAnnotations
+    .map((a: JobAnnotation) => a.message)
+    .join(";");
```
Comment (tcsullens) [inline src/tracing/job.ts:228]: Joining several failure messages together isn't ideal, but currently don't see a better way to ensure we're capturing any/all failures (there can be more than one). This still gives us the ability to create an SLI based on matching specific keywords to this string.

**Comment by mrobinson513** [comment]
@tcsullens when able could you pls update this repository so that:
- platform team are all admins
- main branch is protected

Much thanks

**Comment by tcsullens** [comment]
> @tcsullens when able could you pls update this repository so that:
> 
> * platform team are all admins
> * main branch is protected
> 
> Much thanks

@brett-anderson Do you have permissions to help out here?

File: src/tracing/job.ts (hunk 1 above)
Comment (ekala) [reply src/tracing/job.ts:178]: 1. For the infra-related failures, is it possible to perform additional attribution e.g. which failures originated from GitHub vs. our own infrastructure?
1. Based on the commentary (in docstrings) above, does this mean that we have an infrastructure failure when `github.conclusion == failure` and `github.job.annotations is not null`?

File: src/tracing/job.ts (hunk 2 above)
Comment (ekala) [reply src/tracing/job.ts:228]: Unfortunately, Honeycomb doesn't have support for arrays. As for capturing these failure messages and possibly using these to created an SLI (based on a derived field), we can feature flag this using the [ci-experiments action](https://github.com/wrapbook/ci-experiments-action) to gather some data on a subset of the failures

Comment (tcsullens) [reply src/tracing/job.ts:228]: >we can feature flag this using the ci-experiments action to gather some data on a subset of the failures

We can do that!

File: src/tracing/job.ts (hunk 1 above)
Comment (tcsullens) [reply src/tracing/job.ts:178]: >For the infra-related failures, is it possible to perform additional attribution e.g. which failures originated from GitHub vs. our own infrastructure?

This is the hope! However, the messages in these annotations vary and I think we'll need to gather some data and work to classify distinct messages into buckets, eg. the PR description example, `Could not assume role with OIDC...` I think we'd classify as a `infra`-type failure. 

>Based on the commentary (in docstrings) above, does this mean that we have an infrastructure failure when github.conclusion == failure and github.job.annotations is not null?

Following on my clarification above, that's not necessarily the case as we'd need to inspect the annotation message to really determine root cause; the comments/logic here is just to exclude capturing _other_ types of Job failures, like a `rspec` failure — does that make sense?

**Review by brett-anderson — APPROVED** [review]
Looks good! Thanks @tcsullens!

File: src/tracing/job.ts (hunk 1 above)
Comment (tcsullens) [reply src/tracing/job.ts:178]: @ekala Following up here, I was able to collect a sample of annotation messages from our recent GHA jobs [here](https://www.notion.so/wrapbook/Add-annotations-to-CI-traces-126fdd08fcdc80cba708d82d9d8a9a66?pvs=4#131fdd08fcdc80ae9ac4ce6381d05af1) — I think we'll be able to attribute some of these as specifically `infra` or `github`-type errors

**Pull request merged**